src/openlifu/db/database.py

```diff
@@ -107,6 +107,35 @@ def write_session(self, subject:Subject, session:Session, on_conflict=OnConflict
 
         self.logger.info(f"Added session with ID {session_id} for subject {subject.id} to the database.")
 
+    def write_run(self, subject_id:str, session_id:str, run:Run, on_conflict=OnConflictOpts.ERROR):
+        # Check whether the run already exist in the session
+        run_ids = self.get_run_ids(subject_id, session_id)
+
+        if run.id in run_ids:
+            if on_conflict == OnConflictOpts.ERROR:
+                raise ValueError(f"Run with ID {run.id} already exists for session {session_id}.")
+            elif on_conflict == OnConflictOpts.OVERWRITE:
+                raise ValueError("OVERWRITE is not supported for write_run")
+            elif on_conflict == OnConflictOpts.SKIP:
+                self.logger.info(f"Skipping run with ID {run.id} for session {session_id} as it already exists.")
+                return  # Skip adding the session
+            else:
+                raise ValueError("Invalid 'on_conflict' option. Use 'error', 'overwrite', or 'skip'.")
+
+        # Save the run metadata to a JSON file and copy volume data file to database
+        run_metadata_filepath = self.get_run_metadata_filepath(subject_id, session_id, run.id)
+        Path(run_metadata_filepath).parent.mkdir(exist_ok=True)
+        run.to_file(run_metadata_filepath)
+
+        # Write snapshot of the session
+        subject = self.load_subject(subject_id)
+        session = self.load_session(subject, session_id)
+        session.to_file(run_metadata_filepath.parent / f'{run.id}_session_snapshot.json')
+
+        # Write snapshot of the protocol
+        protocol = self.load_protocol(session.protocol_id)
+        protocol.to_file(run_metadata_filepath.parent / f'{run.id}_protocol_snapshot.json')
```

Comment:
I agree, in that case, i think it would be better to give only session as input with a bool argument write_snapshot. Protocol and Subject could be extracted from "session". For this approach, i want to make sure that even when the session is not saved yet, it points to a protocol and a subject that are saved ? or it could points to a subject and a protocol that are not saved yet.   
def write_run(self, session:Session, run:Run, write_snapshot=False, on_conflict=OnConflictOpts.ERROR):

Reply:
@ebrahimebrahim I agree with (1) and (2) but when the user does not provide a protocol as input of write_run, should we still write a snapshot of the session ? if yes, then should we write with it the protocol for session.protocol_id ?

Reply:
I think it's the same with protocol: only write a protocol snapshot if the user provides it, and if the user does not provide it then don't write it.